File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.cc
#include <iostream>
using namespace std;

int n, s;
int arr[20];
int cnt = 0;

void backtracking(int num, int start, int level, int result) {
	if (num == level) {
		if (result == s) {
			cnt++;
		}
		return;
	}

	for (int i = start; i < n; i++) {
		int sum = result;
		sum += arr[i];
		backtracking(num, i + 1, level + 1, sum);
	}
}

void subset_num() {
	for (int i = 1; i <= n; i++) {
		backtracking(i, 0, 0, 0);
	}
}

int main() {
	cin >> n >> s;

	for (int i = 0; i < n; i++) {
		cin >> arr[i];
	}

	subset_num();

	cout << cnt;
}

File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.py
n, s = map(int, input().split())
arr = list(map(int, input().split()))
answer = 0

def f(num, start, level, result):
    global answer
    if num == level:
        if result == s:
            answer += 1
        return
    
    for i in range(start, n):
        tmp_result = result
        tmp_result += arr[i]
        f(num, i + 1, level + 1, tmp_result)

for i in range(1, n + 1):
    f(i, 0, 0, 0)

print(answer)